libraries/discriminator/src/lib.rs

```diff
@@ -21,34 +21,81 @@ mod tests {
 
     #[allow(dead_code)]
     #[derive(SplDiscriminate)]
-    #[discriminator_hash_input("some_discriminator_hash_input")]
+    #[discriminator_hash_input("my_first_instruction")]
     pub struct MyInstruction1 {
         arg1: String,
         arg2: u8,
     }
 
     #[allow(dead_code)]
     #[derive(SplDiscriminate)]
-    #[discriminator_hash_input("yet_another_discriminator_hash_input")]
-    pub struct MyInstruction2 {
-        arg1: u64,
+    #[discriminator_hash_input("global:my_second_instruction")]
+    pub enum MyInstruction2 {
+        One,
+        Two,
+        Three,
     }
 
     #[allow(dead_code)]
     #[derive(SplDiscriminate)]
-    #[discriminator_hash_input("global:my_instruction_3")]
-    pub enum MyInstruction3 {
-        One,
-        Two,
-        Three,
+    #[discriminator_hash_input("global:my_instruction_with_lifetime")]
+    pub struct MyInstruction3<'a> {
+        data: &'a [u8],
+    }
+
+    #[allow(dead_code)]
+    #[derive(SplDiscriminate)]
+    #[discriminator_hash_input("global:my_instruction_with_one_generic")]
+    pub struct MyInstruction4<T> {
+        data: T,
+    }
+
+    #[allow(dead_code)]
+    #[derive(SplDiscriminate)]
+    #[discriminator_hash_input("global:my_instruction_with_one_generic_and_lifetime")]
+    pub struct MyInstruction5<'b, T> {
+        data: &'b [T],
+    }
+
+    #[allow(dead_code)]
+    #[derive(SplDiscriminate)]
+    #[discriminator_hash_input("global:my_instruction_with_multiple_generics_and_lifetime")]
+    pub struct MyInstruction6<'c, U, V> {
+        data1: &'c [U],
+        data2: &'c [V],
+    }
+
+    #[allow(dead_code)]
+    #[derive(SplDiscriminate)]
+    #[discriminator_hash_input(
+        "global:my_instruction_with_multiple_generics_and_lifetime_and_where"
+    )]
+    pub struct MyInstruction7<'c, U, V>
+    where
+        U: Clone + Copy,
+        V: Clone + Copy,
+    {
+        data1: &'c [U],
+        data2: &'c [V],
     }
 
     fn assert_discriminator<T: spl_discriminator::discriminator::SplDiscriminate>(
         hash_input: &str,
+        case_num: u8,
```

Comment:
nit: this `case_num` will be very brittle to maintain -- why not just print `hash_input` on error?